### PR TITLE
fix: select text when opening find/replace

### LINKED ISSUE #1486: Marimo auto-markdown view modifies original code
### Describe the bug

My cell with `mo.md` got automatically turned into a markdown cell, and when I turned it back into python it lost the formatting.

Original input:

```
mo.md(f"""
Hello
""")
```

Code in the cell after transformed back to python from the markdown cell:

```
mo.md(f"Hello")
```

### Environment

0.6.6

### Code to reproduce

```
mo.md(f"""
Hello
""")
```

### LINKED ISSUE #1506: Find and replace should start with text highlighted
The marimo find and replace (started with Ctrl/Cmd+F) should start with the previously entered text selected, just like Chrome's find and replace. That way the user can hit `Ctrl+F` and immediately start typing their query, without having to first select or delete the previous query.

### PR DESCRIPTION
**Pull request by metaboulie**
## 📝 Summary

Fixes #1506

## 📋 Checklist

- [x] I have read the [contributor guidelines](../CONTRIBUTING.md).
- [x] For large changes, or changes that affect the public API: this change was discussed or approved through an issue, on [Discord](https://discord.gg/JE7nhX6mD8), or the community [discussions](https://github.com/marimo-team/marimo/discussions) (Please provide a link if applicable).
- [x] I have added tests for the changes made.
- [x] I have run the code and verified that it works as expected.

## 📜 Reviewers

@akshayka OR @mscolnick

### CODE REVIEW

File: frontend/src/core/codemirror/language/markdown.ts [hunk 1]
```diff
@@ -159,7 +163,10 @@ function splitQuotePrefix(quote: string): [PrefixKind, string] {
   );
   for (const prefix of prefixKindsByLength) {
     if (quote.startsWith(prefix)) {
-      return [prefix, quote.slice(prefix.length)];
+      const remaining = quote.slice(prefix.length);
+      if (remaining.startsWith('"""') || remaining.startsWith('"')) {
```
Comment (mscolnick) [inline frontend/src/core/codemirror/language/markdown.ts:167]: do we want this? i would expect `'''` and `'` to be valid as well?

File: frontend/src/core/codemirror/language/markdown.ts [hunk 2]
```diff
@@ -38,13 +38,16 @@ const regexes = pairs.map(
     ] as const,
 );
 
+type QuoteType = '"' | '"""'; // Define a type that restricts to either single or triple quotes
```
Comment (mscolnick) [inline frontend/src/core/codemirror/language/markdown.ts:41]: i think `'` and `'''` are valid as well

Comment (mscolnick) [reply frontend/src/core/codemirror/language/markdown.ts:41]: if its not too much trouble, could we add some tests to `markdown.test.ts`?

```python
  describe("should not modify original code when going in and out", () => {
    const cases = [
      'mo.md("""# Markdown Title\n\nSome content here.""")',
      'mo.md("Some *markdown* content")',
      "mo.md('''# Another Title\nContent''')",
      'mo.md(f"""This is some \\"""content\\"""!""")',
      'mo.md("""This is some \\"""content\\"""!""")',
      'mo.md("This is some \\"content\\"!")',
      "mo.md('This is some \\'content\\'!')",
      'mo.md("""Markdown with an escaped \\"quote\\"""")',
      'mo.md("""# Title\n\n```python\nprint("Hello, Markdown!")\n```""")',
      'mo.md("")',
      'mo.md("# Title without proper delimiters")',
      'mo.md("""   \n# Title\nContent\n   """)',
    ];
    it.each(cases)("should not modify %s", (pythonCode) => {
      const [innerCode] = adapter.transformIn(pythonCode);
      const [wrappedCode] = adapter.transformOut(innerCode);
      expect(pythonCode).toBe(wrappedCode);
    });
  });
```

there is also a case where we cannot keep the `QuoteType`. like when we go from single-line to multi-line, we need to upgrade from single-quote to multi-quote. we should make sure that is handled

File: frontend/src/components/find-replace/find-replace.tsx [hunk 3]
```diff
@@ -52,6 +53,13 @@ export const FindReplace: React.FC = () => {
     return openFindReplacePanel();
   });
 
+  useEffect(() => {
+    if (state.isOpen && findInputRef.current) {
+      findInputRef.current.focus(); // Focus the input
+      findInputRef.current.select(); // Select all text in the input
+    }
+  }, [state.isOpen]); // Depend on isOpen to trigger when the panel opens
```
Comment (mscolnick) [inline frontend/src/components/find-replace/find-replace.tsx:61]: awesome! nice touch

File: frontend/src/core/codemirror/language/markdown.ts (hunk 2 above)
Comment (mscolnick) [reply frontend/src/core/codemirror/language/markdown.ts:41]: basically - i don't think we want to downgrade from a `"""` to `"` it if was previously a `"""`. but we do want to upgrade to `"""`, if the markdown becomes multi-line

Comment (metaboulie) [reply frontend/src/core/codemirror/language/markdown.ts:41]: > basically - i don't think we want to downgrade from a `"""` to `"` it if was previously a `"""`. but we do want to upgrade to `"""`, if the markdown becomes multi-line

yeah, this is what i intend to do

Comment (metaboulie) [reply frontend/src/core/codemirror/language/markdown.ts:41]: > i think `'` and `'''` are valid as well

i should've considered these two cases

Comment (metaboulie) [reply frontend/src/core/codemirror/language/markdown.ts:41]: > if its not too much trouble, could we add some tests to `markdown.test.ts`?
> 
> ```python
>   describe("should not modify original code when going in and out", () => {
>     const cases = [
>       'mo.md("""# Markdown Title\n\nSome content here.""")',
>       'mo.md("Some *markdown* content")',
>       "mo.md('''# Another Title\nContent''')",
>       'mo.md(f"""This is some \\"""content\\"""!""")',
>       'mo.md("""This is some \\"""content\\"""!""")',
>       'mo.md("This is some \\"content\\"!")',
>       "mo.md('This is some \\'content\\'!')",
>       'mo.md("""Markdown with an escaped \\"quote\\"""")',
>       'mo.md("""# Title\n\n```python\nprint("Hello, Markdown!")\n```""")',
>       'mo.md("")',
>       'mo.md("# Title without proper delimiters")',
>       'mo.md("""   \n# Title\nContent\n   """)',
>     ];
>     it.each(cases)("should not modify %s", (pythonCode) => {
>       const [innerCode] = adapter.transformIn(pythonCode);
>       const [wrappedCode] = adapter.transformOut(innerCode);
>       expect(pythonCode).toBe(wrappedCode);
>     });
>   });
> ```
> 
> there is also a case where we cannot keep the `QuoteType`. like when we go from single-line to multi-line, we need to upgrade from single-quote to multi-quote. we should make sure that is handled

yeah, i'll add more tests

**Review by mscolnick — APPROVED** [review]
awesome! thanks for the PR

**Pull request merged**